### PR TITLE
fix: type QueryRenderer environment prop as IEnvironment

### PR DESCRIPTION
**Pull request by jdk243**
Update `QueryRendererProps` so that `environment` is typed as `IEnvironment`, matching the [flow definitions in relay](https://github.com/facebook/relay/blob/master/packages/react-relay/ReactRelayQueryRenderer.js#L60). This then allows passing the `environment` property from `ReactRelayContext` as the value for the `environment` prop on `QueryRenderer`s

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/facebook/relay/blob/master/packages/react-relay/ReactRelayQueryRenderer.js#L60>
- ~[ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~
- ~[ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.~

### CODE REVIEW

**Comment by alloy** [comment]
It’s unclear to me what the wish is here, we already have the environment prop typed. Perhaps a PR will better explain what your suggested change is?

**Comment by alloy** [comment]
I’m still a little unsure about the change, are you just changing the import so as to exactly follow the upstream codebase or do you actually expect something else to happen? If so, can you add test coverage.

**Comment by jdk243** [comment]
@alloy ah I'm sorry, I'll update the description. 

I was trying to solve an issue for a use case where I was retrieving `environment` from `ReactRelayContext` in a component via `useContext` so that I could then pass it as the environment prop to `QueryRenderer`. The type of the `environment` property on `ReactRelayContext` is that `IEnvironment` type, exported from [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/relay-runtime/index.d.ts#L61) in `relay-runtime`. However, the `environment` prop for `QueryRender` was typed as `RelayModernEnvironment` from `relay-runtime`, [see here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/relay-runtime/index.d.ts#L130). This caused a TS issue when trying to pass RelayContext's environment as  the value for the environment prop to a `QueryRenderer`. Looking at the flow definitions for QueryRenderer [here](https://github.com/facebook/relay/blob/master/packages/react-relay/ReactRelayQueryRenderer.js#L60) it seemed like the intention was to have the environment prop have the `IEnvironment` type, not the  `RelayModernEnvironment` implementation of the interface.

Did that all make sense?

File: types/react-relay/test/react-relay-tests.tsx [hunk 1]
```diff
@@ -714,3 +714,45 @@ requestSubscription(
...
+
+const MyRelayContextConsumer: React.FunctionComponent = () => {
+    const relayContext = React.useContext(ReactRelayContext);
+    if (!relayContext || !relayContext.environment) {
+        return null;
+    }
+
+    return (
+        <QueryRenderer<MyQuery>
+            environment={relayContext.environment}
```
Comment (alloy) [inline types/react-relay/test/react-relay-tests.tsx]: So to be clear, this test failed perviously, yeah?

**Review by alloy — COMMENTED** [review]
Ahh, I see–yeah I get it now, so we were exporting a concrete implementation of the interface, whereas it should be the abstract interface? Makes sense 👍

File: types/react-relay/test/react-relay-tests.tsx (hunk 1 above)
Comment (jdk243) [reply types/react-relay/test/react-relay-tests.tsx]: yep! if the change in the type defs is reverted this will start throwing an error

Comment (alloy) [reply types/react-relay/test/react-relay-tests.tsx]: Ace, thanks for confirming 👍

**Comment by alloy** [comment]
Great job, thanks! 🙏

**Comment by rubas** [comment]
Maybe this is related as [this](https://github.com/jdk243/DefinitelyTyped/blob/bf2463814ffc8e5333cf510928eb85b75dd9df6b/types/react-relay/test/react-relay-tests.tsx#L61) fails for me now with `Type 'RelayModernEnvironment' is not assignable to type 'Environment'.`

 "@types/react-relay": "^7.0.2",
 "@types/relay-runtime": "^6.0.11",

**Comment by jdk243** [comment]
@rubas I'm not seeing any issue there, which version of typescript are you using?

**Comment by Seriously-Bot-KiK** [comment]
> @rubas I'm not seeing any issue there, which version of typescript are you using?

With TS 3.7.2 and 3.8.0-dev.

This error came up in our project and I was able to replicate it with your test file.

```
TS2769: No overload matches this call.
  Overload 1 of 2, '(props: Readonly<{ cacheConfig?: CacheConfig | null | undefined; fetchPolicy?: "network-only" | "store-and-network" | undefined; } & QueryRendererProps<OperationType>>): ReactRelayQueryRenderer<...>', gave the following error.

    Type 'RelayModernEnvironment' is not assignable to type 'Environment'.

      Types of property 'applyUpdate' are incompatible.
        Type '(optimisticUpdate: import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction) => import("./node_modules/@types/relay-runtime/lib/util/RelayRuntimeTypes").Disposable' is not assignable to type '(optimisticUpdate: import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction) => import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib...'.

          Types of parameters 'optimisticUpdate' and 'optimisticUpdate' are incompatible.
            Type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction' is not assignable to type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction'.

              Types of property 'storeUpdater' are incompatible.
                Type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").StoreUpdater' is not assignable to type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").StoreUpdater'.

                  Types of parameters 'store' and 'store' are incompatible.
                    Type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").RecordSourceProxy' is not assignable to type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").RecordSourceProxy'.

                      The types of 'create(...).copyFieldsFrom' are incompatible between these types.
                        Type '(source: RecordProxy<{}>) => void' is not assignable to type '(source: RecordProxy) => void'.

                          Types of parameters 'source' and 'source' are incompatible.
                            Type 'RecordProxy' is not assignable to type 'RecordProxy<{}>'.

                              The types returned by 'getLinkedRecord(...)' are incompatible between these types.
                                Type 'RecordProxy | null | undefined' is not assignable to type 'RecordProxy<any>'.
                                  Type 'undefined' is not assignable to type 'RecordProxy<any>'.

  Overload 2 of 2, '(props: { cacheConfig?: CacheConfig | null | undefined; fetchPolicy?: "network-only" | "store-and-network" | undefined; } & QueryRendererProps<OperationType>, context?: any): ReactRelayQueryRenderer<...>', gave the following error.
    Type 'RelayModernEnvironment' is not assignable to type 'Environment'.
```

**Comment by rubas** [comment]
> @rubas I'm not seeing any issue there, which version of typescript are you using?

With TS 3.7.2 and 3.8.0-dev.

This error came up in our project and I was able to replicate it with your test file.

```
TS2769: No overload matches this call.
  Overload 1 of 2, '(props: Readonly<{ cacheConfig?: CacheConfig | null | undefined; fetchPolicy?: "network-only" | "store-and-network" | undefined; } & QueryRendererProps<OperationType>>): ReactRelayQueryRenderer<...>', gave the following error.

Type 'RelayModernEnvironment' is not assignable to type 'Environment'.

Types of property 'applyUpdate' are incompatible.
Type '(optimisticUpdate: import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction) => import("./node_modules/@types/relay-runtime/lib/util/RelayRuntimeTypes").Disposable' is not assignable to type '(optimisticUpdate: import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction) => import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib...'.

Types of parameters 'optimisticUpdate' and 'optimisticUpdate' are incompatible.
Type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction' is not assignable to type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction'.

Types of property 'storeUpdater' are incompatible.
Type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").StoreUpdater' is not assignable to type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").StoreUpdater'.

Types of parameters 'store' and 'store' are incompatible.
Type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").RecordSourceProxy' is not assignable to type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").RecordSourceProxy'.

The types of 'create(...).copyFieldsFrom' are incompatible between these types.
Type '(source: RecordProxy<{}>) => void' is not assignable to type '(source: RecordProxy) => void'.

Types of parameters 'source' and 'source' are incompatible.
 Type 'RecordProxy' is not assignable to type 'RecordProxy<{}>'.

The types returned by 'getLinkedRecord(...)' are incompatible between these types.
 Type 'RecordProxy | null | undefined' is not assignable to type 'RecordProxy<any>'.
Type 'undefined' is not assignable to type 'RecordProxy<any>'.

 Overload 2 of 2, '(props: { cacheConfig?: CacheConfig | null | undefined; fetchPolicy?: "network-only" | "store-and-network" | undefined; } & QueryRendererProps<OperationType>, context?: any): ReactRelayQueryRenderer<...>', gave the following error.
    Type 'RelayModernEnvironment' is not assignable to type 'Environment'.
```

**Comment by alloy** [comment]
> Type 'import("./node_modules/@types/react-relay/node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction' is not assignable to type 'import("./node_modules/@types/relay-runtime/lib/store/RelayStoreTypes").OptimisticUpdateFunction'.

Looks like you have multiple copies of `@types/relay-runtime` installed, try to get rid of the one installed in `@types/react-relay`.

**Comment by rubas** [comment]
> Looks like you have multiple copies of `@types/relay-runtime` installed, try to get rid of the one installed in `@types/react-relay`.

True 🙄 

I had to delete `yarn.lock`, clear `node_modules` and `yarn install` to get ride of ` node_modules/@types/react-relay/node_modules...`. This fixed my issue. Thanks for help.

**Comment by renanmav** [comment]
This "multiple copies of `@types/relay-runtime`" seems to be a recurrent problem.

**Pull request merged**